plugins/rtp/src/device.vala

```diff
@@ -400,16 +400,41 @@ public class Dino.Plugins.Rtp.Device : MediaDevice, Object {
             }
             if (best_index == -1) {
                 // No caps in first round, try without framerate
+                int target_pixel_count = 1280 * 720;
+                double target_pixel_ratio = 16.0 / 9.0;
+                int best_pixel_count = 0;
+                double best_pixel_ratio = 0.0;
+                double best_ratio_penalty = 0.0;
+
                 for (int i = 0; i < device.caps.get_size(); i++) {
                     unowned Gst.Structure? that = device.caps.get_structure(i);
                     if (!that.has_name("video/x-raw")) continue;
                     int width = 0, height = 0;
-                    if (!that.has_field("width") || !that.get_int("width", out width)) continue;
-                    if (!that.has_field("height") || !that.get_int("height", out height)) continue;
-                    if (best_width < width || best_width == width && best_height < height) {
+                    if (!that.has_field("width") || !that.get_int("width", out width) || width == 0) continue;
+                    if (!that.has_field("height") || !that.get_int("height", out height) || height == 0) continue;
+
+                    int pixel_count = width * height;
+                    double pixel_ratio;
+                    if (width > height)
+                        pixel_ratio = (double)width / height;
+                    else
+                        pixel_ratio = (double)height / width;
+
+                    double ratio_penalty;
+                    if (pixel_ratio > target_pixel_ratio)
+                        ratio_penalty = pixel_ratio / target_pixel_ratio;
+                    else
+                        ratio_penalty = target_pixel_ratio / pixel_ratio;
+
+                    if (best_width == 0 ||
+                         (ratio_penalty * (pixel_count - target_pixel_count).abs()) <
```

Comment:
This gives a resolution a value of 0 if it has 1280*720 pixels, no matter the ratio. While I agree that resolutions like 960x960, 1024x900, 1152x800, 1200x768, 1440x640, 1536x600 or 1600x576 are probably not that common to appear on a device that also provides any other resolution that's 1280*720 pixels. You could do the same that you did for `ratio_penalty` and calculate a `size_penalty` (that turns 1 for 1280x720 pixels instead of 0).

Then you could calculate `cost = ratio_penalty * size_penalty` and only store `best_cost` rather than `best_ratio_penalty`, `best_pixel_count` and `best_pixel_ratio`.

Reply:
Very good point, thanks! Did that and it's not just cleaner and more correct - the factors also looks much more intuitive. Just used `diversion_factor` instead of `penalty` or `cost`.